btree/btree.c

```diff
@@ -2939,145 +2953,128 @@ static void ff_capture(struct slot *slot, struct m0_be_tx *tx)
+ * will add those credits to @cred.
+ */
+static void ff_node_alloc_credit(const struct nd *node,
+				 struct m0_be_tx_credit *cred)
 {
-	struct ff_head *h           = ff_data(node);
-	int             shift       = h->ff_shift;
-	int             node_size   = 1ULL << shift;
-	int             header_size = sizeof(*h);
-
-	m0_be_allocator_credit(NULL, M0_BAO_FREE_ALIGNED,
-			       node_size, shift, accum);
+	int         shift     = ff_shift(node);
+	m0_bcount_t node_size = 1ULL << shift;
 
-	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(1, header_size));
+	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED,
+			       node_size, shift, cred);
 }
-#ifndef __KERNEL__
+
 /**
- *  --------------------------------------------
- *  Section START - Btree Credit
- *  --------------------------------------------
+ * This function will calculate credits required to update the node and it will
+ * add those credits to @cred.
  */
-/**
- * Credit for btree put and delete operation.
- * For put operation, at max 2 nodes can get captured in each level plus an
- * extra node. The delete operation can use less nodes, still use the same api
- * to be on the safer side.
- *
- * @param accum transaction credit.
- */
-static void btree_callback_credit(struct m0_be_tx_credit *accum)
+static void ff_node_update_credit(const struct nd *node,
+				   struct m0_be_tx_credit *cred)
 {
-	accum->tc_cb_nr += BTREE_CALLBACK_CREDIT;
+	int         shift     = ff_shift(node);
+	m0_bcount_t node_size = 1ULL << shift;
+
+	m0_be_tx_credit_add(cred, &M0_BE_TX_CREDIT(1, node_size));
 }
 
 /**
- * This function will calculate credits required to allocate node and it will
- * add those credits to @accum.
+ * This function will calculate credits required to free the node and it will
+ *  add those credits to @cred.
  */
-static void btree_node_alloc_credit(const struct m0_btree  *tree,
-				    struct m0_be_tx_credit *accum)
+static void ff_node_free_credit(const struct nd *node,
+				struct m0_be_tx_credit *cred)
 {
-	m0_bcount_t             node_size;
-	int                     shift;
+	struct ff_head *h           = ff_data(node);
+	int             shift       = h->ff_shift;
+	int             node_size   = 1ULL << shift;
+	int             header_size = sizeof(*h);
 
-	shift     = node_shift(tree->t_desc->t_root);
-	node_size =  1ULL << shift;
+	m0_be_allocator_credit(NULL, M0_BAO_FREE_ALIGNED,
+			       node_size, shift, cred);
 
-	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED,
-			       node_size, shift, accum);
+	m0_be_tx_credit_add(cred, &M0_BE_TX_CREDIT(1, header_size));
 }
 
 /**
- * This function will calculate credits required to update node and it will add
+ * This function will calculate credits required to split node and it will add
  * those credits to @accum.
  */
-static void btree_node_update_credit(const struct m0_btree  *tree,
-				     struct m0_be_tx_credit *accum)
+static void ff_node_split_credit(const struct nd *node,
+				 struct m0_be_tx_credit *cred)
```

Comment:
We do not need to have a ff_node_split_credit() because nodes themselves do not know about splitting. The concept of splitting is done above the node layer. So the higher layer *_tick() functions should do the split credit calculations if needed using the node credit functions already provided.

Reply:
Made changes as per the discussion.